### PR TITLE
Mitigate JSON writing performance regression

### LINKED ISSUE #595: Possible JSON logging performance regression
Keith reported to me a possible loss of performance in the JSON logging of Zeek 3.0.0 versus Bro 2.6.4, so I did a quick benchmark using this `test.zeek`:

```zeek
option log_mult = 0;

event connection_state_remove(c: connection) &priority=-10
  {
  local n = log_mult;

  while ( n > 0 )
    {
    Log::write(Conn::LOG, c$conn);
    --n;
    }
  }
```

```
# Zeek 3.0.0 JSON logging
time -f "%e %M" zeek -G random.seed -r 2009-M57-day11-18.trace test.zeek log_mult=100 LogAscii::use_json=T
45.00 946668

# Bro 2.6.4 JSON logging
time -f "%e %M"  bro -G random.seed -r 2009-M57-day11-18.trace test.zeek log_mult=100 LogAscii::use_json=T
15.79 140820

# Zeek 3.0.0 ASCII logging
time -f "%e %M" zeek -G random.seed -r 2009-M57-day11-18.trace test.zeek log_mult=100 LogAscii::use_json=F
14.29 156476

# Bro 2.6.4 ASCII logging
time -f "%e %M"  bro -G random.seed -r 2009-M57-day11-18.trace test.zeek log_mult=100 LogAscii::use_json=F
15.77 149264

# Zeek 3.0.0 JSON logging, patched to use default nlohmann json object instead of fifo_map
time -f "%e %M" zeek -G random.seed -r 2009-M57-day11-18.trace test.zeek log_mult=100 LogAscii::use_json=T
21.96 615240
```

And some flamegraphs:

[bro-2.6.4-json.zip](https://github.com/zeek/zeek/files/3649495/bro-2.6.4-json.zip)

11.79% spent in `threading::formatter::JSON::Describe()`

[zeek-3.0.0-json-fifo-map.zip](https://github.com/zeek/zeek/files/3649498/zeek-3.0.0-json-fifo-map.zip)

63.52% spent in `threading::formatter::JSON::Describe()`

[zeek-3.0.0-json-default.zip](https://github.com/zeek/zeek/files/3649497/zeek-3.0.0-json-default.zip) (not using fifo_map)

44.72% spent in `threading::formatter::JSON::Describe()`

So at least a good portion of the time relates to the ordering of JSON output.  Maybe making that a Zeek configuration option would be something to partially help the loss of performance here.

@timwoj other ideas for things that could help here?  Is there anything related to "correctness" of the JSON output in Zeek 3.0.0 versus that of Bro 2.6.4 that helps justify some of the difference ?  Or is it clearly just a performance difference between two json-formatting implementations (ours vs. nlohmann) ?

it's also possible my test case is too contrived and hitting some new worst-case scenario we never really expect, but doubt it.

### PR DESCRIPTION
**Pull request by timwoj**
Fixes #595 

Requires https://github.com/zeek/zeek-3rdparty/pull/3 to be merged

The times below are using the test script and command line from the GHI linked above.

On `master`:

```
real	0m21.563s
user	0m25.880s
sys	0m5.525s
```

With changes on this branch:

```
real	0m13.034s
user	0m17.429s
sys	0m5.499s
```

Bro 2.6.4:
```
real	0m12.878s
user	0m16.419s
sys	0m6.537s
```

### CODE REVIEW

**Comment by timwoj** [comment]
Added bro 2.6.4 metrics.

File: src/Val.cc [hunk 1]
```diff
@@ -646,9 +646,8 @@ static ZeekJson BuildJSON(Val* val, bool only_loggable=false, RE_Matcher* re=new
 
 		case TYPE_OPAQUE:
 			{
-			j = ZeekJson::object();
 			auto* oval = val->AsOpaqueVal();
-			j["opaque_type"] = OpaqueMgr::mgr()->TypeID(oval);
+			j = {"opaque_type", OpaqueMgr::mgr()->TypeID(oval)};
```
Comment (jsiwek) [inline src/Val.cc]: I expected a map-like initializer to look like `{ {key, value} }`, not `{key, value}`, so thought this change might actually be making an array now instead of an object ?  This example seems to confirm:

```
ZeekJson j;
j = { {"one", "two"} };
printf("%s\n", j.dump().data());
j = {"one", "two"} ;
printf("%s\n", j.dump().data());
```

Outputs:

```
{"one":"two"}
["one","two"]
```

File: src/threading/formatters/JSON.cc [hunk 2]
```diff
@@ -182,11 +182,7 @@ ZeekJson JSON::BuildJSON(Value* val, const string& name) const
 		}
 
 	if ( ! name.empty() && ! j.is_null() )
-		{
-		ZeekJson j2 = ZeekJson::object();
-		j2[name] = j;
-		return j2;
-		}
+		return { name, j };
```
Comment (jsiwek) [inline src/threading/formatters/JSON.cc]: This makes an array, not an object now?

**Comment by jsiwek** [comment]
Also, the license in `COPYING.3rdparty` should be updated (remove fifo_map, add tsl).

File: src/Val.cc (hunk 1 above)
Comment (timwoj) [reply src/Val.cc]: Yep, you're right. I misread the docs. I think there's also a missing unit test there since none of them caught that.

**Comment by timwoj** [comment]
Fixed the above comments. I added a unit test for the change in Val.cc. The block in JSON.cc that changed is never actually called, but it's correct now. I made a note to go back and look at whether that block can just be removed.

**Comment by rsmmr** [comment]
Can we get this evaluated in one of the environments that had performance problems with the 3.0.0 release to see if it addresses the problem sufficiently? Note we should also include any other JSON changes (thinking of #611), so maybe just start merging it all into the 3.0 branch?

**Comment by sethhall** [comment]
I'll start pushing on that.

**Comment by jsiwek** [comment]
I'll work on getting this merged along with other things into the `release/3.0` branch to make it easier for people to test (and after, I'll also ping Keith directly to help test).
